File: projects/evaldesk/evals/schema_probe.py
"""Prepare offline validation from verified original output, without model calls."""
import json
from pathlib import Path
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from engine import encode
from schemas import SCHEMAS
from store import load_evidence

ROOT = Path(__file__).resolve().parents[1]
rows = []
for project, profile in [('feedbacklens', 'json-feedback-v1'), ('changelens', 'json-change-v1')]:
    evidence = load_evidence(ROOT / 'artifacts' / (project + '-regression-run'))
    for cell in evidence['cells']:
        rows.append({'id': project + '/' + cell['id'], 'profile': profile, 'output': cell['output'],
                     'run_id': evidence['manifest']['id'], 'original_status': cell['status']})

feedback = {'suggestions': [{'theme': '导出错误', 'kind': 'problem', 'quote': '导出一直报错'}]}
change = {'summary': '新增导出审计记录。', 'old_quote': '', 'new_quote': '新增审计记录。'}
fixtures = [
    ('feedback-valid', 'json-feedback-v1', feedback, True),
    ('feedback-empty', 'json-feedback-v1', {'suggestions': []}, True),
    ('feedback-array', 'json-feedback-v1', [feedback], False),
    ('feedback-missing', 'json-feedback-v1', {'suggestions': [{'theme': '错误'}]}, False),
    ('feedback-kind', 'json-feedback-v1', {'suggestions': [{**feedback['suggestions'][0], 'kind': 'billing'}]}, False),
    ('feedback-extra', 'json-feedback-v1', {**feedback, 'approved': True}, False),
    ('feedback-item-extra', 'json-feedback-v1', {'suggestions': [{**feedback['suggestions'][0], 'priority': 1}]}, False),
    ('feedback-too-many', 'json-feedback-v1', {'suggestions': feedback['suggestions'] * 6}, False),
    ('feedback-short', 'json-feedback-v1', {'suggestions': [{**feedback['suggestions'][0], 'theme': '错'}]}, False),
    ('change-valid', 'json-change-v1', change, True),
    ('change-missing', 'json-change-v1', {'summary': '新增导出审计记录。'}, False),
    ('change-null', 'json-change-v1', {**change, 'old_quote': None}, False),
    ('change-short', 'json-change-v1', {**change, 'summary': '变化'}, False),
    ('change-extra', 'json-change-v1', {**change, 'approved': True}, False),
    ('change-long', 'json-change-v1', {**change, 'new_quote': '字' * 4001}, False),
    # Semantically false but structurally valid: do not promise meaning validation.
    ('structure-not-truth', 'json-change-v1', {**change, 'summary': '完全无依据的业务结论。'}, True),
]
for identity, profile, output, expected in fixtures:
    rows.append({'id': identity, 'profile': profile, 'output': json.dumps(output, ensure_ascii=False), 'expected': expected})
rows.append({'id': 'fence-preserved', 'profile': 'json-change-v1',
             'output': '```json\n' + json.dumps(change) + '\n```', 'expected': False})
target = ROOT / 'data/schema-probe-input.json'
target.parent.mkdir(exist_ok=True)
target.write_bytes(encode({'schemas': SCHEMAS, 'rows': rows}))
print(target)

File: projects/evaldesk/evals/schema_probe.mjs
// Offline assertion replay, using the same pinned engine as live evaluations.
process.env.PROMPTFOO_DISABLE_TELEMETRY = '1';
process.env.PROMPTFOO_DISABLE_UPDATE = '1';
process.env.PROMPTFOO_DISABLE_REMOTE_GENERATION = 'true';
const fs = await import('node:fs/promises');
const {assertions} = await import('promptfoo');
const input = JSON.parse(await fs.readFile(new URL('../data/schema-probe-input.json', import.meta.url), 'utf8'));
const pkg = JSON.parse(await fs.readFile(new URL('../node_modules/promptfoo/package.json', import.meta.url), 'utf8'));
if (pkg.version !== '0.123.1') throw Error('Pinned engine required');
const rows = [];
for (const row of input.rows) {
  const options = {test: {}, providerResponse: {output: row.output}};
  const syntax = await assertions.runAssertion({...options, assertion: {type:'is-json'}});
  const structure = await assertions.runAssertion({...options, assertion: {type:'is-json', value: input.schemas[row.profile]}});
  if (row.expected !== undefined && structure.pass !== row.expected) throw Error('Unexpected validation: '+row.id);
  rows.push({...row, syntax: {pass:syntax.pass, reason:syntax.reason}, structure: {pass:structure.pass, reason:structure.reason}});
}
const output = {scope:'Offline replay of unchanged model output plus synthetic boundary fixtures; zero new model calls', engine:pkg.version, schemas:input.schemas, rows};
await fs.writeFile(new URL('../artifacts/schema-probe.json',import.meta.url), JSON.stringify(output,null,2));
console.log(JSON.stringify({rows:rows.length,fixtures:rows.filter(r=>r.expected!==undefined).length,newlyRejected:rows.filter(r=>r.run_id && r.syntax.pass&&!r.structure.pass).map(r=>r.id)}));
